File: drills/elixir/solutions/nc74_longest_increasing_subsequence__patience.exs
defmodule Solution do
  # Patience sorting. Keep the smallest value that any subsequence of each
  # length ends with; that list is always sorted, so each number either extends
  # it or replaces the first entry it is no bigger than. The list is not the
  # answer subsequence -- only its length is meaningful.
  def length_of_lis(nums) do
    nums
    |> Enum.reduce([], &place(&2, &1, []))
    |> length()
  end

  defp place([], n, seen), do: Enum.reverse([n | seen])

  defp place([first | rest], n, seen) do
    if first >= n,
      do: Enum.reverse([n | seen]) ++ rest,
      else: place(rest, n, [first | seen])
  end
end

File: drills/go/solutions/nc74_longest_increasing_subsequence__patience.go
package main

import "sort"

func lengthOfLIS(nums []int) int {
	// tails[k] is the smallest tail of any increasing subsequence of
	// length k+1. Each number replaces the first tail not below it (or
	// extends the list); the list's length is the answer. O(n log n).
	tails := []int{}
	for _, n := range nums {
		i := sort.SearchInts(tails, n)
		if i == len(tails) {
			tails = append(tails, n)
		} else {
			tails[i] = n
		}
	}
	return len(tails)
}
